### PR TITLE
HKG: Car Port for Kia Niro EV 2023 (HDA1, European version)

### PR DESCRIPTION
**Pull request by JosselinLecocq**
**Checklist**
- [X] added entry to CarInfo in selfdrive/car/*/values.py and ran `selfdrive/car/docs.py` to generate new docs
- [X] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [X] route with openpilot: b153671049a867b3|2023-03-31--13-16-13
- [X] route with stock system: b153671049a867b3|2023-03-19--14-49-19
- [X] harness type: [Hyundai A](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)
- [X] architecture: CAN-FD

### CODE REVIEW

File: selfdrive/car/tests/routes.py [hunk 1]
```diff
@@ -129,6 +129,7 @@
   CarTestRoute("007d5e4ad9f86d13|2021-09-30--15-09-23", HYUNDAI.KIA_K5_2021),
   CarTestRoute("c58dfc9fc16590e0|2023-01-14--13-51-48", HYUNDAI.KIA_K5_HEV_2020),
   CarTestRoute("50c6c9b85fd1ff03|2020-10-26--17-56-06", HYUNDAI.KIA_NIRO_EV),
+  CarTestRoute("b153671049a867b3|2023-03-31--13-16-13", HYUNDAI.KIA_NIRO_EV_2ND_GEN),
```
Comment (sshane) [inline selfdrive/car/tests/routes.py:132]: It looks like the logs for this route are gone, can you provide a new one with rlogs uploaded? (you can click the upload button for "All logs" in connect)

File: selfdrive/car/hyundai/values.py [hunk 2]
```diff
@@ -1304,6 +1306,19 @@ class Buttons:
       b'\xf1\x00DEE MFC  AT KOR LHD 1.00 1.03 95740-Q4000 180821',
     ],
   },
+  CAR.KIA_NIRO_EV_2ND_GEN: {
+    (Ecu.fwdRadar, 0x7D0, None): [
+      b'\xf1\x00SG2_ RDR -----      1.00 1.01 99110-AT000         ',
+      b'\xf1\x8799110AT000\xf1\x00SG2_ RDR -----      1.00 1.01 99110-AT000         ',
```
Comment (sshane) [inline selfdrive/car/hyundai/values.py:1312]:
```suggestion
```

File: selfdrive/car/hyundai/values.py [hunk 3]
```diff
@@ -1304,6 +1306,19 @@ class Buttons:
...
+    ],
+    (Ecu.eps, 0x7D4, None): [
+      b'\xf1\x00SG2 MDPS C A.00 1.03 56310AO000\x00 4SGVC103',
+      b'\xf1\x8756310AO000\x00\xf1\x00SG2 MDPS C A.00 1.03 56310AO000\x00 4SGVC103',
+    ],
```
Comment (sshane) [inline selfdrive/car/hyundai/values.py:1317]:
```suggestion
```

File: selfdrive/car/tests/routes.py (hunk 1 above)
Comment (JosselinLecocq) [reply selfdrive/car/tests/routes.py:132]: Can you try again ? I just clicked on the upload button for "All logs" on this route.

File: selfdrive/car/hyundai/values.py (hunk 3 above)
Comment (JosselinLecocq) [reply selfdrive/car/hyundai/values.py]: For my understanding, why are you deleting the "eps" section ?

File: selfdrive/car/tests/routes.py (hunk 1 above)
Comment (sshane) [reply selfdrive/car/tests/routes.py:132]: They're still missing. It looks you may have a smaller sized C3 variant, which won't store as many logs. I would try to get a route then quickly upload to WiFi after

Comment (JosselinLecocq) [reply selfdrive/car/tests/routes.py:132]: Ok, check this one instead : b153671049a867b3|2023-04-05--10-00-30

File: selfdrive/car/hyundai/interface.py [hunk 4]
```diff
@@ -136,10 +136,15 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long):
...
+    elif candidate in (CAR.KIA_NIRO_EV, CAR.KIA_NIRO_EV_2ND_GEN, CAR.KIA_NIRO_PHEV, CAR.KIA_NIRO_HEV_2021, CAR.KIA_NIRO_HEV_2ND_GEN):
       ret.mass = 3452. * CV.LB_TO_KG + STD_CARGO_KG  # average of all the cars
       ret.wheelbase = 2.7
-      ret.steerRatio = 13.9 if CAR.KIA_NIRO_HEV_2021 else 13.73  # Spec
+      if candidate == CAR.KIA_NIRO_EV_2ND_GEN:
+        ret.steerRatio = 13.3 # Spec : https://www.kiamedia.com/us/en/models/niro-ev/2023/specifications
+      elif candidate == CAR.KIA_NIRO_HEV_2021:
+        ret.steerRatio = 13.9 # Spec
+      else:
+        ret.steerRatio = 13.73 # Spec
```
Comment (adeebshihadeh) [inline selfdrive/car/hyundai/interface.py:147]: @nuwandavek do these differences even matter? seems like we can just put the average in here

**Comment by sshane** [comment]
@JosselinLecocq can you get another route with the stock system and upload that for debugging purposes? Not a blocker but would be useful to have.

File: selfdrive/car/tests/routes.py (hunk 1 above)
Comment (sshane) [reply selfdrive/car/tests/routes.py:132]: That works, thanks!

File: selfdrive/car/torque_data/params.yaml [hunk 5]
```diff
@@ -44,6 +44,7 @@ JEEP GRAND CHEROKEE V6 2018: [2.27116, 1.4057367824262523, 0.11725947414922003]
 KIA EV6 2022: [3.2, 2.093457, 0.05]
 KIA K5 2021: [2.405339728085138, 1.460032270828705, 0.11650989850813716]
 KIA NIRO EV 2020: [2.9215954981365337, 2.1500583840260044, 0.09236802474810267]
+KIA NIRO EV 2023: [2.9215954981365337, 2.5, 0.09236802474810267]
```
Comment (sshane) [inline selfdrive/car/torque_data/params.yaml:47]: If you just copied these, we should add this to the substitute file

File: selfdrive/car/hyundai/interface.py (hunk 4 above)
Comment (sshane) [reply selfdrive/car/hyundai/interface.py:147]: the learned ratio is 14,.05 so I doubt it makes much of a difference

File: selfdrive/car/hyundai/interface.py [hunk 6]
```diff
@@ -136,10 +136,10 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long):
       ret.mass = 1985. + STD_CARGO_KG
       ret.wheelbase = 2.78
       ret.steerRatio = 14.4 * 1.1   # 10% higher at the center seems reasonable
-    elif candidate in (CAR.KIA_NIRO_EV, CAR.KIA_NIRO_PHEV, CAR.KIA_NIRO_HEV_2021, CAR.KIA_NIRO_HEV_2ND_GEN):
+    elif candidate in (CAR.KIA_NIRO_EV, CAR.KIA_NIRO_EV_2ND_GEN, CAR.KIA_NIRO_PHEV, CAR.KIA_NIRO_HEV_2021, CAR.KIA_NIRO_HEV_2ND_GEN):
       ret.mass = 3452. * CV.LB_TO_KG + STD_CARGO_KG  # average of all the cars
       ret.wheelbase = 2.7
-      ret.steerRatio = 13.9 if CAR.KIA_NIRO_HEV_2021 else 13.73  # Spec
+      ret.steerRatio = 13.6  # average of the platforms
```
Comment (sshane) [inline selfdrive/car/hyundai/interface.py:142]:
```suggestion
      ret.steerRatio = 13.6  # average of all the cars
```

File: selfdrive/car/hyundai/interface.py [hunk 7]
```diff
@@ -136,10 +136,10 @@ def _get_params(ret, candidate, fingerprint, car_fw, experimental_long):
       ret.mass = 1985. + STD_CARGO_KG
       ret.wheelbase = 2.78
       ret.steerRatio = 14.4 * 1.1   # 10% higher at the center seems reasonable
-    elif candidate in (CAR.KIA_NIRO_EV, CAR.KIA_NIRO_PHEV, CAR.KIA_NIRO_HEV_2021, CAR.KIA_NIRO_HEV_2ND_GEN):
+    elif candidate in (CAR.KIA_NIRO_EV, CAR.KIA_NIRO_EV_2ND_GEN, CAR.KIA_NIRO_PHEV, CAR.KIA_NIRO_HEV_2021, CAR.KIA_NIRO_HEV_2ND_GEN):
       ret.mass = 3452. * CV.LB_TO_KG + STD_CARGO_KG  # average of all the cars
```
Comment (sshane) [inline selfdrive/car/hyundai/interface.py:140]:
```suggestion
      ret.mass = 3543. * CV.LB_TO_KG + STD_CARGO_KG  # average of all the cars
```

**Pull request merged**